file.go

```diff
@@ -7,56 +7,109 @@ import (
 )
 
 var (
-	ErrNotDirectory = errors.New("Couldn't call NextFile(), this isn't a directory")
-	ErrNotReader    = errors.New("This file is a directory, can't use Reader functions")
+	ErrNotDirectory = errors.New("file isn't a directory")
+	ErrNotReader    = errors.New("file isn't a regular file")
+
+	ErrNotSupported = errors.New("operation not supported")
 )
 
 // File is an interface that provides functionality for handling
 // files/directories as values that can be supplied to commands. For
-// directories, child files are accessed serially by calling `NextFile()`.
+// directories, child files are accessed serially by calling `NextFile()`
+//
+// Read/Seek methods are only valid for files
+// NextFile method is only valid for directories
 type File interface {
-	// Files implement ReadCloser, but can only be read from or closed if
-	// they are not directories
-	io.ReadCloser
-
-	// FileName returns a filename associated with this file
-	FileName() string
-
-	// FullPath returns the full path used when adding with this file
-	FullPath() string
-
-	// IsDirectory returns true if the File is a directory (and therefore
-	// supports calling `NextFile`) and false if the File is a normal file
-	// (and therefor supports calling `Read` and `Close`)
-	IsDirectory() bool
-
-	// NextFile returns the next child file available (if the File is a
-	// directory). It will return (nil, io.EOF) if no more files are
-	// available. If the file is a regular file (not a directory), NextFile
-	// will return a non-nil error.
-	NextFile() (File, error)
+	io.Closer
+
+	// Size returns size of this file (if this file is a directory, total size of
+	// all files stored in the tree should be returned). Some implementations may
+	// choose not to implement this
+	Size() (int64, error)
 }
 
-type StatFile interface {
+// Regular represents a regular Unix file
+type Regular interface {
 	File
 
-	Stat() os.FileInfo
+	io.Reader
+	io.Seeker
+}
+
+// DirEntry exposes information about a directory entry
+type DirEntry interface {
+	// Name returns the base name of this entry, which is the base name of
+	// the referenced file
+	Name() string
+
+	// File returns the file referenced by this DirEntry
+	File() File
+
+	// Regular is an alias for ent.File().(Regular). If the file isn't a regular
+	// file, nil value will be returned
+	Regular() Regular
+
+	// Dir is an alias for ent.File().(directory). If the file isn't a directory,
+	// nil value will be returned
+	Dir() Directory
 }
 
-type PeekFile interface {
-	SizeFile
+// DirIterator is a iterator over directory entries.
+// See Directory.Entries for more
+type DirIterator interface {
+	// DirEntry holds information about current directory entry.
+	// Note that after creating new iterator you MUST call Next() at least once
+	// before accessing these methods. Calling these methods without prior calls
+	// to Next() and after Next() returned false may result in undefined behavior
```

Comment:
I actually think a separate DirIterator type may be an overkill but @Stebalien may disagree.  It kind of creates a mismatch between `Regular` that can be modified by seeking and `Directory`.

My original proposal for an iterator was not well though out and I apologize if I misleading you.

After giving this some thought I think it would be better to just have a `NextFile()` method in directory that returns a `DirEntry` i.e.:
```
type Directory {
  io.Closer
  Size() (int64, error)
  NextFile() (DirEntry, error)
}
```
espacally since we may only want to support iterating over a directory once.

Also by returning the `DirEntry` with the `Next()` or `NextFile()` method you avoid the need for this long explanation.

Reply:
The part I *like* about modifying it in-place is that we make it clear that it can't be reused. However, I don't really have many strong opinions one way or another.

Reply:
So I've tinkered with these and few other approaches quite a bit beyond this. The current (`DirIterator`) way is by far the nicest to use as the API consumer.

For usecases where it's not true, the user probably doesn't really need to iterate over the directory, we could provide a `Lookup(name string) File` method. 

I actually quite like the fact that having separate iterator decouples state from Directory object, allowing murtiple iterations at one time, but I agree it smells funny. We could just move these mothods to directory struct and provide `Clone` method on File which would be a bit like unix `dup(2)` except that it would not share state.

Reply:
So I've tinkered with these and few other approaches quite a bit beyond this. The current (`DirIterator`) way is by far the nicest to use as the API consumer.

For usecases where it's not true, the user probably doesn't really need to iterate over the directory, we could provide a `Lookup(name string) File` method. 

I actually quite like the fact that having separate iterator decouples state from Directory object, allowing murtiple iterations at one time, but I agree it smells funny. We could just move these mothods to directory struct and provide `Clone` method on File which would be a bit like unix `dup(2)` except that it would not share state.

Reply:
Yeah, if we are going to have an separate iterator it should not effect the internal state of the Directory.  If it does, its better than the Directory entry itself is the iterator.  This is more aligned with how `readdir()` works in the c-lib.

Reply:
> it should not effect the internal state of the Directory

This should be the case for most implementations (not really doable for multipartfile, but we can return `nil` if user calls `Entries` more than once), so that's probably fine